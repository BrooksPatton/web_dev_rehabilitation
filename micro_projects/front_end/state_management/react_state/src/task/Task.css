.Task {
    width: 15rem;
    display: flex;
    justify-content: space-between;
}

.Task-delete {
    background-color: firebrick;
    border: 0;
    color: white;
}

.Task-delete:hover {
    background-color: indianred;
}

.Task-edit {
    background-color: inherit;
    border: 1px solid lightslategray;
}

.Task-edit:hover {
    background-color: darksalmon;
}