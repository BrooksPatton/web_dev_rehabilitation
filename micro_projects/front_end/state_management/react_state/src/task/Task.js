import React from 'react';
import './Task.css';

class Task extends React.Component {
    constructor(properties) {
        super(properties);

        this.state = {
            task: this.props.task,
            editing: false,
            editedTask: this.props.task.name
        };
    }

    renderEditing = () => (
        <form onSubmit={this.handleSaveChangeToTask}>
            <label htmlFor={`Task-edit-${this.state.task.id}`}></label>
            <input type="text" value={this.state.editedTask} id={`Task-edit-${this.state.task.id}`} onChange={this.handleEditTaskChange}></input>
        </form>
    )

    handleEditTaskChange = event => this.setState({ editedTask: event.target.value })

    renderEditButton = () => (
        <button className="Task-edit" onClick={() => this.setState({ editing: true })}>Edit</button>
    )

    renderSaveButton = () => (
        <button className="Task-edit" onClick={this.handleSaveChangeToTask}>Save</button>
    )

    handleSaveChangeToTask = event => {
        event.preventDefault();

        console.log('handled');


        const { id } = this.state.task;

        this.props.editTaskName({ id, name: this.state.editedTask });
        this.setState({ editing: false })
    }

    renderCancelEditButton = () => (
        <button className="Task-edit" onClick={this.cancelEdit}>Cancel</button>
    )

    cancelEdit = () => {
        this.setState({
            editing: false,
            editedTask: this.state.task.name
        });
    }

    toggleCompletedStatus = () => this.props.toggleTaskCompletedStatus(this.state.task.id)

    renderTaskName = () => (
        <span className={this.state.task.completed ? 'Task-completed' : null}>{this.state.task.name}</span>
    )

    removeTask = () => this.props.removeTaskById(this.state.task.id);

    render() {
        return (
            <section className="Task">
                <div className="Task-name">
                    <label htmlFor={`${this.state.task.id}-task`}></label>
                    <input type="checkbox" id={`${this.state.task.id}-task`} onChange={this.toggleCompletedStatus} checked={this.state.task.completed}></input>
                    {this.state.editing ? this.renderEditing() : this.renderTaskName()}
                </div>
                <div>
                    {this.state.editing ? this.renderSaveButton() : this.renderEditButton()}
                    {this.state.editing ? this.renderCancelEditButton() : null}
                    <button className="Task-delete" onClick={this.removeTask}>X</button>
                </div>
            </section>
        )
    }
}

export default Task;