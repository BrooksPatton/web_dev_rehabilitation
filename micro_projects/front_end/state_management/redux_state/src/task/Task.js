import React from 'react';
import { connect } from 'react-redux';
import './Task.css';
import { editTaskName, toggleTaskCompletion, deleteTask, updateEditedTaskName, toggleEditedStatus, saveEditedTaskName } from '../store/store';

function TaskComponent({ task, updateEditedTaskName, toggleTaskCompletion, toggleEditedStatus, saveEditedTaskName, deleteTask }) {
    const renderEditing = () => (
        <form onSubmit={handleSaveEditedTaskName}>
            <label htmlFor={`Task-edit-${task.id}`}></label>
            <input type="text" value={task.editedName} id={`Task-edit-${task.id}`} onChange={event => updateEditedTaskName({ id: task.id, editedName: event.target.value })}></input>
        </form>
    )

    const handleSaveEditedTaskName = event => {

        event.preventDefault();
        saveEditedTaskName(task.id);
    }

    const renderEditButton = () => (
        <button className="Task-edit" onClick={() => toggleEditedStatus(task.id)}>Edit</button>
    )

    const renderSaveButton = () => (
        <button className="Task-edit" onClick={handleSaveEditedTaskName}>Save</button>
    )

    const renderCancelEditButton = () => (
        <button className="Task-edit" onClick={() => toggleEditedStatus(task.id)}>Cancel</button>
    )

    const toggleCompletedStatus = () => toggleTaskCompletion(task.id)

    const renderTaskName = () => (
        <span className={task.completed ? 'Task-completed' : null}>{task.name}</span>
    )

    return (
        <section className="Task">
            <div className="Task-name">
                <label htmlFor={`${task.id}-task`}></label>
                <input type="checkbox" id={`${task.id}-task`} checked={task.completed} onChange={toggleCompletedStatus}></input>
                {task.editing ? renderEditing() : renderTaskName()}
            </div>
            <div>
                {task.editing ? renderSaveButton() : renderEditButton()}
                {task.editing ? renderCancelEditButton() : null}
                <button className="Task-delete" onClick={() => deleteTask(task.id)}>X</button>
            </div>
        </section>
    )
}

const Task = connect(null, { editTaskName, toggleTaskCompletion, deleteTask, updateEditedTaskName, toggleEditedStatus, saveEditedTaskName })(TaskComponent);

export default Task;