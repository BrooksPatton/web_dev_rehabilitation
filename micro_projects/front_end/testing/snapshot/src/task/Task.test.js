import React from 'react';
import renderer from 'react-test-renderer';

import Task from './Task';

it('should render the Task component with a task', () => {
    const task = {
        id: 1,
        name: 'I am a task name',
        completed: false
    };

    const functionMock = () => { };

    const tree = renderer
        .create(<Task task={task} editTaskName={functionMock} toggleTaskCompletedStatus={functionMock} removeTaskById={functionMock} />)
        .toJSON();

    expect(tree).toMatchSnapshot();
});

it('should render the Task component with a completed task', () => {
    const task = {
        id: 1,
        name: 'I am a completed task',
        completed: true
    };

    const functionMock = () => { };

    const tree = renderer
        .create(<Task task={task} editTaskName={functionMock} toggleTaskCompletedStatus={functionMock} removeTaskById={functionMock} />)
        .toJSON();

    expect(tree).toMatchSnapshot();
});