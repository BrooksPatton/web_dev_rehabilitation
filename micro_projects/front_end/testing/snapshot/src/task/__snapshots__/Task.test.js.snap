// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the Task component with a completed task 1`] = `
<section
  className="Task"
>
  <div
    className="Task-name"
  >
    <label
      htmlFor="1-task"
    />
    <input
      checked={true}
      id="1-task"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="Task-completed"
    >
      I am a completed task
    </span>
  </div>
  <div>
    <button
      className="Task-edit"
      onClick={[Function]}
    >
      Edit
    </button>
    <button
      className="Task-delete"
      onClick={[Function]}
    >
      X
    </button>
  </div>
</section>
`;

exports[`should render the Task component with a task 1`] = `
<section
  className="Task"
>
  <div
    className="Task-name"
  >
    <label
      htmlFor="1-task"
    />
    <input
      checked={false}
      id="1-task"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className={null}
    >
      I am a task name
    </span>
  </div>
  <div>
    <button
      className="Task-edit"
      onClick={[Function]}
    >
      Edit
    </button>
    <button
      className="Task-delete"
      onClick={[Function]}
    >
      X
    </button>
  </div>
</section>
`;
